### PR TITLE
Bonnie v1.3 object equivalence equality and datetime comparison updates

### PR DESCRIPTION
**Pull request by hossenlopp**
Pull requests into cql-execution require the following.
Submitter and reviewer should ✔ when done.
For items that are not-applicable, mark "N/A" and ✔.

[CDS Connect](https://cds.ahrq.gov/cdsconnect) and [Bonnie](https://github.com/projecttacoma/bonnie) are the main users of this repository. 
It is strongly recommended to include a person from each of those projects as a reviewer.

**Submitter:**
- [x] This pull request describes why these changes were made https://jira.mitre.org/browse/BONNIE-1702
- [x] Code diff has been done and been reviewed (it does not contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass
- [x] Code coverage has not gone down and all code touched or added is covered.
- [x] ~~All dependent libraries are appropriately updated or have a corresponding PR related to this change~~ Will do later

**Reviewer:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

File: src/util/comparison.coffee [hunk 1]
```diff
@@ -83,6 +83,8 @@ deepCompareKeysAndValues = (a, b, comparisonFunction) ->
   # Array.every() will only return true or false, so set a flag for if we should return null
   shouldReturnNull = false
   finalComparisonResult = aKeys.length is bKeys.length and aKeys.every (key) ->
+    # if both are null we should return true to satsify ignoring empty values in tuples
```
Comment (losborne) [inline src/util/comparison.coffee]: satsify -> satisfy

File: src/datatypes/datetime.coffee [hunk 2]
```diff
@@ -676,6 +570,243 @@ DateTime.prototype.isSamePrecision = Date.prototype.isSamePrecision = (other) ->
...
+DateTime.prototype.before = Date.prototype.before = (other, precision) ->
+  if not((other.isDate) or (other.isDateTime))
+    return null
+  else if @.isDate and other.isDateTime
+    return @getDateTime().before(other, precision)
+  else if @.isDateTime and other.isDate
+    other = other.getDateTime()
+
+  if precision? && @constructor.FIELDS.indexOf(precision) < 0
+    throw new Error("Invalid precision: #{precision}")
```
Comment (losborne) [inline src/datatypes/datetime.coffee]: We could also enumerate the valid precisions in this error message, which might be helpful.

File: test/datatypes/interval.coffee [hunk 3]
```diff
@@ -739,11 +739,19 @@ describe 'DateTimeInterval.union', ->
 
     [x, y] = xy @dIvl.before
     # TODO: I don't know about these tests... doesn't make sense to me.
```
Comment (losborne) [inline test/datatypes/interval.coffee]: I think this TODO might be able to be removed

File: src/datatypes/datetime.coffee (hunk 2 above)
Comment (hossenlopp) [reply src/datatypes/datetime.coffee]: The CQL to ELM translator rejects invalid precisions. The only case that we will see this is if someone uploads a CQL 1.2 measure that uses `week` precision. Now that I think about it, I will create a test to make sure the right list is being used to reject the time level precisions on Date objects.

**Pull request merged**